### DISCUSSION TITLE
Next Gen stubs

### DISCUSSION THREAD

**Discussion by CEXT-Dan** [CEXT-Dan/PyRx#12] [Ideas]
I made a prototype of function overloading. Need to implement it.
![overloads](https://github.com/CEXT-Dan/PyRx/assets/26830697/f1216836-c813-492d-9e14-a784b8943b6d)

**Comment by CEXT-Dan** [comment]
should be...
Additional Args:
or  
Overloads:

**Comment by gswifort** [comment]
Meybe `@typing.overload`

**Comment by CEXT-Dan** [comment]
That’s the one! 

I have to write them all by hand, so it’s going to take a minute.
VS Code won’t read what boost::python emits. There’s not a whole lot of them
Hoping to get the constructors first and commonly used functions first

**Comment by gswifort** [comment]
I don't know how boost::python works, but I see that many methods have stubs - were they generated automatically by `boost::python` or by you?
Do you have any influence on how methods are generated in `*.pyi` or is everything automatically generated?
If it's not possible to customize how `boost::python` works, you can probably always parse the generated `*.pyi` file and add docstring-based methods:

```
class ArcDimension:
    def __init__ (self, *args, **kwargs)-> None :
      '''__init__( (object)arg1) -> None :

    C++ signature :
        void __init__(struct _object * __ptr64)

__init__( (object)arg1) -> None :

    C++ signature :
        void __init__(struct _object * __ptr64)

__init__( (object)arg1, (ObjectId)arg2) -> None :

    C++ signature :
        void __init__(struct _object * __ptr64,class PyDbObjectId)

__init__( (object)arg1, (ObjectId)arg2, (OpenMode)arg3) -> None :

    C++ signature :
        void __init__(struct _object * __ptr64,class PyDbObjectId,enum AcDb::OpenMode)'''

##########################

class ArcDimension:
    @overload
    def __init__(self, /) -> None : ...

    @overload
    def __init__(self, arg1: ObjectId, /) -> None : ...
    
    @overload
    def __init__(self, arg1: ObjectId, arg2: OpenMode, /) -> None : ...

    def __init__ (self, *args, **kwargs)-> None : ...
```

the docstrings do not provide the names of the method parameters, I assume it is not arg1, arg2, arg3 etc., that's why I used positional-only parameters

**Comment by CEXT-Dan** [comment]
I have a stub generator, what you posted generated from boost’s docstring, without modification
I’ve been inserting named parameters into the docstring to generate a more correct stub

I just learned about the @overload, so I’m doing exactly what you suggest, parsing that docstring
I insert the named parameters into the docstring from the C++ side. Then parse it. 

I’ll go through and do all the constructors first.

this is the output of Db.Text.... getting close : )

```python
class Text:

    @overload
    def __init__ (self, /)-> None :
        ...

    @overload
    def __init__ (self, position: PyGe.Point3d, text: str)-> None :
        ...

    @overload
    def __init__ (self, position: PyGe.Point3d, text: str, height: float, rotation: float)-> None :
        ...

    @overload
    def __init__ (self, id: PyDb.ObjectId)-> None :
        ...

    @overload
    def __init__ (self, id: PyDb.ObjectId, mode: PyDb.OpenMode)-> None :
        ...

    @overload
    def __init__ (self, id: PyDb.ObjectId, mode: PyDb.OpenMode, erased: bool)-> None :
        ...
    def __init__ (self)-> None :
      '''Overloads:
    - None: Any
    - position: PyGe.Point3d, text: str
    - position: PyGe.Point3d, text: str, height: float, rotation: float
    - id: PyDb.ObjectId
    - id: PyDb.ObjectId, mode: PyDb.OpenMode
    - id: PyDb.ObjectId, mode: PyDb.OpenMode, erased: bool
    '''
    ...
```

**Comment by gswifort** [comment]
I see that in each class you define all methods, including inherited ones:

```py
class Entity(DbObject):
    def addReactor (self, reactor: PyDb.EntityReactor)-> None :
        ...

class Curve(Entity):
    def addReactor (self, reactor: PyDb.EntityReactor)-> None :
        ...

class Line(Curve):
    def addReactor (self, reactor: PyDb.EntityReactor)-> None :
        ...
```

it is enough only in the first class:

```py
class Entity(DbObject):
    def addReactor (self, reactor: PyDb.EntityReactor)-> None :
        ...

class Curve(Entity):
        ...

class Line(Curve):
        ...
```
![image](https://github.com/CEXT-Dan/PyRx/assets/131259064/4766ebc5-4a9a-462c-9c2d-e7557b6ad14f)

**Reply by CEXT-Dan** [reply]
Yes, but not on purpose. The doc string generator is a total hack, I started rolling it when I knew zero python and now it’s morphed onto a mess lol

Using inspect 
for each class in module
--for each method in class 
----make docstring from the C++ signature 

Gives every method. A small number are actually virtual overrides from a C++ perspective.
This will definitely be one of the areas that's worked on when I refactor 

is it causing an issue?

**Reply by gswifort** [reply]
It does not cause problems, but it is good to know where a given method is defined - whether it is the same for all `Entity` or, for example, the `Line` class has a different implementation.
I understand you are talking about the [genDoc.py](https://github.com/CEXT-Dan/PyRx/blob/main/PyRxStubs/utility/genDoc.py) module? Maybe I can suggest solutions.

**Reply by CEXT-Dan** [reply]
I figured it was a side affect of boost Boost generates a docstring like”
```
“setDatabaseDefaults( (Entity)arg1) -> None :
    C++ signature :
        void setDatabaseDefaults(class PyDbEntity {lvalue})
setDatabaseDefaults( (Entity)arg1, (Database)arg2) -> None :
    ![(self, db: PyDb.Database = 'current')]!<[{4351}]>
    C++ signature :
        void setDatabaseDefaults(class PyDbEntity {lvalue},class PyDbDatabase)”
```
I inject stuff from the C++ side, I.e.

> ![(self, db: PyDb.Database = 'current')]!<[{4351}]>

and parse it

BTW, someone on the boost::python team is supposedly working on compliant stubs.
So I had intended to refactor at some point.

**Reply by gswifort** [reply]
If a given function doesn't have a docstring id assigned (`![(self)]!<[{-1}]>`), can I simply search `DocStrings.json` by the name `Db.BlockTableRecord.blockScaling → acdbblocktablerecord__blockscaling`? I know that wrapper classes aren't always named the same as the original classes, but maybe it's easier to create a file mapping wrapper names to original names than assigning an id to each function?

**Reply by CEXT-Dan** [reply]
-1 means I have not assigned it yet, eventually they will all have IDs assigned. 
I have an SQLIte database (DocString.db) that I use to store the docstrings, then I export it to JSON. The number is the ROWID

Exporting to JSON is a workaround because SQLite wont load into BricsCAD due to a binary conflict and I need to be able to generate stubs from that platform.  I’m considering changing to duckdb and removing the need for the JSON, just make the query

For now, the process is, if an item does not have a docstring, add it to the database, export the table to Jason (SQLiteStudio)
Add the entry to the CPP file 

I.e. from
```C++
.def("blockScaling", &PyDbBlockTableRecord::blockScaling, DS.ARGS())
.def("setBlockScaling", &PyDbBlockTableRecord::setBlockScaling, DS.ARGS({ "val : PyDb.BlockScaling" }))
```

To:
```C++
.def("blockScaling", &PyDbBlockTableRecord::blockScaling, DS.ARGS(XXXX))
.def("setBlockScaling", &PyDbBlockTableRecord::setBlockScaling, DS.ARGS({ "val : PyDb.BlockScaling" },XXXX))
```

If it’s in the JSON file, it’s already in the database, so it just needs the number assigned. 
There’s probably a thousand or so to do… I usually do these when my brain is toast and I just do busy work

**Comment by gswifort** [comment]
Hi, I worked a bit on the stubs.
In my opinion they are almost perfect :)
I have to refactor the generating code a bit, but I'm uploading the generated files.
Please correct a few detected errors:

```raw
PyAp.Document: def setUserData(self, data: any, /): ... 
                                           ^^^

PyEd.Editor: def ssget(mode: str, arg1: any, arg2: any, filter: list = None, /):
                                        ^^^        ^^^

PyPl.PlotFactory: def createPreviewEngine(flags: int = default, /) -> PyPl.PlotEngine: ... 
                                                       ^^^^^^^

PyGi.Geometry: def circularArc(self, p1: PyGe.Point3d, p2: PyGe.Point3d, p3: PyGe.Point3d, arcType: PyGe.ArcType, /)
                                                                                                    ^^^^^^^^^^^^ (PyGe.ArcType ?)
                                                                                                                  
PyGi.Geometry: def polyline(self, vertexList, normal: PyGe.Vector3d = default, marker: int = default, /) -> bool:
                                                                      ^^^^^^^                ^^^^^^^
                                                                                                                  
PyDb.ArcDimension: def __init__( self, centerPoint: PyGe.point3d, xLine1Point: PyGe.Point3d, xLine2Point: PyGe.Point3d, arcPnt: PyGe.Point3d, /): ...
                                                         ^^^^^^^

PyDb.DiametricDimension: def __init__(self, chordPoint: PyGe.point3d, farChordPoint: PyGe.Point3d, leaderLength: float, /): ...
                                                             ^^^^^^^

PyDb.Point3AngularDimension: def __init__(self, centerPoint: PyGe.point3d, xL1Start: PyGe.Point3d, xL1End: PyGe.Point3d, arcPnt: PyGe.Point3d, /): ...
                                                                  ^^^^^^^

PyDb.BlockReference: def appendAttribute(self, val: PyDb.Attribute, /) -> PyDb.ObjectId:
                                                    ^^^^^^^^^^^^^^ (AttributeReference)

PyDb.Core: def getReservedString(reservedType: PyDb.reservedStringEnumType, bGetLocalized: bool, /) -> str:
                                               ^^^^^^^^^^^ (PyDb.ReservedStringEnumType)

           def isReservedString(val: str, reservedType: PyDb.reservedStringEnumType, /) -> bool:
                                                        ^^^^^^^^^^^ (PyDb.ReservedStringEnumType)

PyDb.DimStyleTableRecord: def arrowId(self, val: PyDb.DimArrowFlags, /) -> PyDb.ObjectId: ...
                                                 ^^^^^^^^^^^^^^^^^^ ?????

PyDb.Polyline3d: def appendVertex(self, vertex: PyDb.Vertex3d, /): ...
                                                ^^^^^^^^^^^^^ PyDb.Vertex ??

PyDb.Polyline3d: def insertVertexAt(self, indexVt: PyDb.Vertex3d, newVertex: PyDb.Vertex3d, /): ...
                                                   ^^^^^^^^^^^^^ PyDb.Vertex ??


PyDb.Table: def setFieldId(self, row: int, col: int, content: int, id: PyDb.ObjectId, nFlag: PyDb.CellOption, /): ...
                                                                                             ^^^^^^^^^^^^^^^ CellProperty ??

PyDb.Table: def setGridLineStyle(self, row: int, col: int, nGridLineType: PyDb.GridLineTypes, nLineStyle: PyDb.GridLineStyle, /): ...
                                                                          ^^^^^^^^^^^^^^^^^^ PyDb.GridLineType (without "s")

PyDb.Table: def setGridLinetype(self, row: int, col: int, nGridLineType: PyDb.GridLineTypes, idLinetype: PyDb.ObjectId, /): ...
                                                                         ^^^^^^^^^^^^^^^^^^ PyDb.GridLineType (without "s")
PyGs.GsView: def setView (self, pos: PyGe.Point3d, target: PyGe.Point3d, upVector: PyGe.Vector3d, fWidth: float, fHeight: float,  PyGs.Projection = kParallel)-> None :
                                                                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

```
[stubs.zip](https://github.com/user-attachments/files/16999110/stubs.zip)

**Reply by gswifort** [reply]
5ac65beb1b6de66ac267111047b66a816ad4faa3

https://github.com/CEXT-Dan/PyRx/blob/main/PyRxCore%2FPyGsView.cpp#L97

argument name is missing

**Reply by CEXT-Dan** [reply]
thanks, I had missed a couple of enums, I've added them

**Reply by gswifort** [reply]
https://github.com/CEXT-Dan/PyRx/blob/main/PyRxCore%2FPyGsView.cpp#L97

PyGs.ViewProjection = PyGs.ViewProjection.kParallel

type hints and default value provided, but argument name missing

**Comment by CEXT-Dan** [comment]
I noticed there's some activity with boost::python's signatures and doc strings. I'll have a look when I get back

**Reply by gswifort** [reply]
Any news?

**Reply by CEXT-Dan** [reply]
No real changes with respect to docstrings, unfortunately. Most projects seem to be using Sphinx , or Doxygen (hand writing the python signatures), and extracting that from the C++ source.  I had tried Sphinx but didn’t get anywhere. I actually think what we have is okay, it’s at least as good or better than PyGrx (gstarcad).

BTW, they’re (gstarcad) using PyBind11, which I had seriously considered, but seems to have the same docstring issues. 
It’s an iterative process, eventually, there will be a need go back and revisit/improve what’s emitted from the C++ side.

**Comment by gswifort** [comment]
If I understand correctly, [``Ge.Point3d.asVector``](https://github.com/CEXT-Dan/PyRx/blob/fe064132df7a8236c680bd0f0f1b0f97c63f2211/PyRxCore/PyAcGe.cpp#L884) docstring is generated automatically. Why is [``Ge.Point3d.x``](https://github.com/CEXT-Dan/PyRx/blob/fe064132df7a8236c680bd0f0f1b0f97c63f2211/PyRxCore/PyAcGe.cpp#L872) docstring not generated?

https://github.com/CEXT-Dan/PyRx/blob/fe064132df7a8236c680bd0f0f1b0f97c63f2211/PyRxCore/PyAcGe.cpp#L868-L884

**Reply by CEXT-Dan** [reply]
I don’t know, I probably need to look at the generator, asVector a ‘.def’, whereas x uses ‘.def_readwrite’ meant to expose data members. 
There’s probably no signature generated since it’s data 

I'll add a string and see of something gets picked up, otherwise, I can add getX() and setX( x : float) methods, if you think it would be helpful 

As a side note, Scale, Point, Vector, and Matrix are the only classes there are directly exposed to Python, unlike all the other classes that use impObj() idiom
That’s why the wrappers are wonky looking

**Reply by gswifort** [reply]
No, property is better. I'm mainly interested in the return type, not the docstring itself.

**Reply by CEXT-Dan** [reply]
Yeah, there’s still work to be done on the generator. X, Y, Z are the data, where they could be properties, like kOrigin is (except non static)
kOrigin isn’t picked up either, which is a bummer, especially with Vector3d. kXAxis, kYAxis, kZAxis.  I’ve been using type hinting 

```Python
        vx : Ge.Vector3d = Ge.Vector3d.kXAxis
        vy : Ge.Vector3d = Ge.Vector3d.kYAxis
        vz : Ge.Vector3d = Ge.Vector3d.kZAxis
        print(vx.crossProduct(vy) == vz)
```

It’s one of the reasons pretty much everything is a method. I had considered adding properties in more places. If for some season I can’t get them to work, at some point, the wrappers will be done and I can go back and hand edit these.

**Reply by CEXT-Dan** [reply]
Just to add, there could be some sort or post generation, where some properties could be injected into the class

**Reply by gswifort** [reply]
I'll add, I'm currently working on improving the stubs. So far I've got this effect:

```py
class Point3d:
# (...) #
    @overload
    def __init__(self, /) -> None: ...
    @overload
    def __init__(self, floats: tuple[float] | list[float], /) -> None: ...
    @overload
    def __init__(self, x: float,y: float,z: float, /) -> None: ...
    @overload
    def __init__(self, pln: PyGe.PlanarEnt, pnt2d: PyGe.Point2d, /) -> None: ...
    def __init__(self, *args) -> None:
        pass
# (...) #
    kOrigin: PyGe.Point3d
# (...) #
    @property
    def x(self, /):
        pass
    @property
    def y(self, /):
        pass
    @property
    def z(self, /):
        pass

```

**Reply by CEXT-Dan** [reply]
Awesome!

**Reply by gswifort** [reply]
Is ``repr`` generated automatically or do you have some influence on it?

According to the [python documentation](https://docs.python.org/3/library/functions.html#repr) the string returned by ``repr`` should return the object it represents after calling ``eval`` on it. 

```py
eval(repr(obj)) == obj
```

For ``Ge.Point3d`` unnecessary ``<>`` are added, which makes generating stubs a bit more difficult.

```py
>>> repr(Ge.Point3d())
<PyGe.Point3d(0.00000000000000,0.00000000000000,0.00000000000000)>
```

**Reply by CEXT-Dan** [reply]
I add that example:
https://github.com/CEXT-Dan/PyRx/blob/eb7ad6f207b91cf188c3803e099bd05dba8093c4/PyRxCore/PyAcGe.cpp#L115-L118

**Reply by CEXT-Dan** [reply]
I can remove the <..>, I don't know why it's there, I thought I got it from an example

**Reply by CEXT-Dan** [reply]
I had only added them to a few items, I.e. Point, Vector, Matrix
I did not add them to reference types

**Reply by gswifort** [reply]
It would be nice if you removed it. I would like to provide values ​​in stubs in addition to types in some places. It can be useful for debugging when, for example, an enum is serialized to int.

**Reply by CEXT-Dan** [reply]
> According to the [python documentation](https://docs.python.org/3/library/functions.html#repr) the string returned by `repr` should return the object it represents after calling `eval` on it.

it also says
>otherwise, the representation is a string enclosed in angle brackets that contains the name of the type of the object together with additional information often including the name and address of the object.

this is where I got the <> from

**Reply by CEXT-Dan** [reply]
So it’s clear, I only had explicitly added repr in a few places, Scale, Point, Vector, Matrix, ObjectId, handle.
I added it to AcCell and TypedValue for lisp, but it’s that same format as toString, basically for debugging. 

I have not added it to enums.

> It would be nice if you removed it.
Do you mean the angle brackets of the whole __repr__ overload?

```
line = Db.Line(Ge.Point3d(0,0,0),Ge.Point3d(100,0,0))
print(repr(line))
```
outputs 
```
<PyDb.Line object at 0x0000020E846F4680>
```

```
eval(repr(obj)) == obj
```

is not clear to me how this works

**Reply by gswifort** [reply]
I want the angle brackets to be removed where the call to ``repr(obj)`` returns a string that when called in python (e.g. in ``PYCMDPROMPT``) will give that object (``eval`` just executes a line of code, e.g. in a REPL). And all default repr containing a type name and memory address should be wrapped in <>

Expected result:

```py
>>> repr(PyDb.Line())
<PyDb.Line object at 0x00000294BBA831F0>
>>> obj1 = PyGe.Point3d()
>>> repr(obj1)
PyGe.Point3d(0.00000000000000,0.00000000000000,0.00000000000000)
>>> obj2 = eval(repr(obj1))
>>> obj3 = PyGe.Point3d(0.00000000000000,0.00000000000000,0.00000000000000)
>>> obj1 == obj2 == obj3
True
```

**Reply by gswifort** [reply]
``obj2 = eval(repr(obj1))`` and ``obj3 = PyGe.Point3d(0.00000000000000,0.00000000000000,0.00000000000000)`` is basically the same call

**Reply by CEXT-Dan** [reply]
Ok, I published a new test build. with the  angle brackets removed

**Comment by gswifort** [comment]
I'm almost done with improving stubs. Please fix these bugs.

![image](https://github.com/user-attachments/assets/cce8a54d-6c25-49a3-9357-b95abf0fcb3a)

**Reply by CEXT-Dan** [reply]
Commit 771363d

**Reply by CEXT-Dan** [reply]
Excellent work! Way better than the hack I had!

**Comment by CEXT-Dan** [comment]
You're like a next level coder, an artist!

looking at PyDb.ErrorStatusException. I see I also have a bug, I'll try to fix this morning

**Comment by gswifort** [comment]
Are all stubs generated on one host? on BRX?

**Reply by CEXT-Dan** [reply]
oh yeah, I forgot about BrxBim

**Reply by gswifort** [reply]
So no? :D I'm asking because we can make a simple launcher (like for testing) that will launch AutoCAD, generate everything available, then launch BricsCad and generate PyBr*

**Reply by CEXT-Dan** [reply]
That would be cool!

**Comment by CEXT-Dan** [comment]
Originally, I had thought about letting the user generate stubs to account for API differences, but then decided not to. Stubs generated from ARX, GRX, ZRX are very close and I can probably do some of the implementation to resolve differences, I.e. AcCmColor
PyBrxBim, PyBrxCv, PyBrx are the only outliers that must be run on BricsCAD.

it could still be an option someday

**Comment by CEXT-Dan** [comment]
Is it possible to generate html docs, like in pydoc, but from the .pyi files instead of the main modules?

**Comment by CEXT-Dan** [comment]
The old generator would warn me if the return type was not fully resolved (I.e. a list or a tuple) and not in return_types.json.
It would just print the function name in the same format as the entry  PyEd::SelectionSet::toList
Is it possible to add this? or it could be added as a test. but if I don't know the entry is missing, I won't know to add it

**Reply by CEXT-Dan** [reply]
or it could be written to gen_pyi.log

**Reply by gswifort** [reply]
I'll add it.

**Comment by CEXT-Dan** [comment]
Pylance failed to parse docstring #6951
https://github.com/microsoft/pylance-release/issues/6951

I guess this could be us?

**Comment by gswifort** [comment]
> https://github.com/CEXT-Dan/PyRx/issues/188#issuecomment-2677177665

I thought about creating `.pyi` files where selected signatures and docstrings will be written manually. Then the pyi file generator will add signatures of missing functions (like it does now). And I think that will be the best place to complete ``Ap.Command`` documentation.

**Reply by gswifort** [reply]
@rdesparbes Do you have any idea how to implement this, or better idea how to combine stubs written directly in pyi files and generated ones?

**Reply by rdesparbes** [reply]
We could look into templating, with [jinja](https://en.wikipedia.org/wiki/Jinja_(template_engine)#Example) for example.

We could write partial .pyi files with `{{ }}` placeholders that would be replaced with:
- Function signatures
- Doctrings
- ...

I'm not sure if I understood everything about your use case, but it could look like:

> PyAx.pyi:
```python
class AcadBlock(PyAx.AcadObject):
    {{ add3DFace }}  # Inserts the signature of the method
    """This is the python docstring of the `add3DFace` method"""
```

**Reply by gswifort** [reply]
I would like to manually create pyi files with definitions of selected classes, methods. I assume that a given function, if defined manually, must be complete (signature + formatted docstring). When generating pyi files from c++ source (as it is now) these functions would be detected and replace the generated ones.
e.g. for the ``Db.Database`` class I would like to overwrite the docstring for ``readDwgFile`` (write signatures, including overload, and docstring) and all other methods would be generated automatically. So jinja templates are not what I expect because there are too many methods to define a placeholder for each one.

**Reply by rdesparbes** [reply]
Maybe we could extract the Abstract Syntax Tree (AST) from the generated .pyi files, browse the symbols, and replace the ones that are defined in the "overwriting" .pyi files ?

One library that does that and that supports .pyi files is [libcst](https://github.com/Instagram/LibCST). I have never used it, and there are probably other candidates, but it looks promising.

**Reply by gswifort** [reply]
#306

**Comment by gswifort** [comment]
For example ``Db.BlockTableRecord`` accepts ``__init__`` without arguments, but the stub generator doesn't see it, how can we solve this? Are always ``__init__`` methods in base classes supported by derived classes too?

![image](https://github.com/user-attachments/assets/210f7bf6-7b29-40e2-b08e-225e6e047bab)

**Reply by CEXT-Dan** [reply]
Yeah, it should behave like `Db.Line`, I forgot about the tables and records. Though the doc string still shows the parent class’s even though I inject the class doc string

`Db.Line` shows the doc string `Db.Curve `

**Comment by CEXT-Dan** [comment]
[asked](https://github.com/microsoft/pylance-release/issues/7303)
